cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+11. Do not use single character names, except for loop indices.
+12. Do not use special characters, except for "_" where allowed.
+13. Do not use "\_" as ﬁrst character. Only use it as the last character for class data member names, not local variable names.
+14. Do not use "\_\_".
+15. Use clear and explanatory variable names.
+
+## 3 -- Style Rules
+1. Do not indent pre-processor directives -- there should be no leading spaces before a directive.  (\*)
+2. Never change the language syntax using `#deﬁne`.
+3. Do not use spaces between a function, constructor, or method name and its opening parenthesis, e.g. `energy()` rather than `energy ()`. A similar style is encouraged but not required when brace initialization is used, e.g. `energy{13}`.
+4. Do not use spaces in front of [] or on either side of -> . For example, `position[i]` instead of `position [i]`.
+5. Separate expressions in a `for` statement by spaces.
+6. Use the same indentation for comments as for the block the comments refer to.
+
+## 4 -- Technical Coding Rules
+1. Protect each header ﬁle from multiple inclusion with:  
+`#ifndef PackageName_SubPackageName_FileName_h`  
+`#define PackageName_SubPackageName_FileName_h`  
+(body of header file)  
+`#endif`  
+If necessary to create a unique name, one can add the directory name:  
+`PackageName_SubPackageName_Directory_FileName_h`.
+2. Each header ﬁle contains one class declaration only. (\*)
+3. Header ﬁles must not contain any implementation except for class templates and code to be inlined.
+4. Do not inline virtual functions.
+5. Do not inline functions which contain control structures which require block scoping.
+6. In your own packages, use forward declarations if they are sufﬁcient.
+7. Do not forward-declare an entity from another package.
+8. Do not use absolute directory names or relative ﬁle paths in `#include` directives.
+9. Use `nullptr`, not "0" or "NULL".
+10. Use types like `int`, `uint32_t`, `size_t`, and `ptrdiff_t` consistently and without mixing them.
+11. Use the `bool` type for booleans.
+12. Copy and move assignment operators should return a reference to `*this`.
+13. For a class, definition of any of the following requires definition of all five: destructor, copy constructor, copy assignment operator, move constructor, and move assignment operator. (\*)
+14. Do not use function-like macros.
+15. Use C++ casts, not C-style casting. (\*)
+16. Do not use the ellipsis notation for function arguments, except for variable argument templates. (\*)
+17. Do not use union types. (\*)
+18. If a class has at least one virtual method, it must have a public virtual destructor or (exceptionally) a protected non-virtual destructor.
+19. When a derived class function overrides a virtual function, always mark it with `override` or `final`.
+20. Pass by value arguments which are not to be modiﬁed and are built-in types or small objects; otherwise pass arguments of class types by reference or, if necessary, by pointer.
```

Comment:
This is actually not the best practice anymore. If a moveable class is being passed as an argument to a member function and the argument will be copied into the class, then the argument should be passed by value.
```C++
    void setValues( std::vector<int> values) { values_ = std::move(values); }
```
This allows one to do
```C++
     thing.setValues({{1}, {2}, {3}});
```
or
```C++
    std::vector<int> v = {{1},{2},{3}};
    thing.setValues(std::move(v));
```
Both of which are more efficient.

Reply:
Chris, for my education, is
```c++
void setValues( std::vector<int> values) { values_ = std::move(values); }
```
just as efficient as
```c++
void setValues( std::vector<int> const& values) { values_ = values; }
void setValues( std::vector<int> && values) { values_ = std::move(values); }
```
?

Reply:
(the latter does seem more efficient code-wise, according to Godbolt with gcc 7.4)

Reply:
The `moveable` is only useful if the value is to be copied internally by the call, else passing by reference is the best.

Reply:
@fwyzard by 'latter' do you mean pass by value or the case where you had two overloaded calls? My test the compiler explorer showed both pieces of code generated the same assembly except for the case where a temporary is passed by non-move, in which case the pass by value was more efficient: https://godbolt.org/z/dxgtlR

Reply:
I see I was running gcc 8.2 not 7.4.

Reply:
@slava77 I agree it is worth to make it more clear, I'm just not sure what such wording should be.

Reply:
@fwyzard By quick look (Godbolt gcc 7.4 `-std=c++17 -O2`; it's possible that my test case was too simple) checking `std::vector<int>` and `std::vector<std::string>` and judging only by the amount of assembly
* if caller does not move, i.e. calls `setValues(vec)`
  * for `int`, pass-by-value has one `lea`+`mov`+`je` more than pass-by-const-ref
  * for `string`, pass-by-const-ref has 60 % more lines than pass-by-value
* if caller does move, i.e. calls `setValues(std::move(vec))`
  * both pass-by-value and pass-by-rvalue-ref give the same assembly for both `int` and `string`

Reply:
The core guidelines recommend

[F.16: For "in" parameters, pass cheaply-copied types by value and others by reference to `const`](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-in)

with
> If the function is going to keep a copy of the argument, in addition to passing by const& (for lvalues), add an overload that passes the parameter by && (for rvalues) and in the body std::moves it to its destination. Essentially this overloads a "will-move-from"; see F.18.   

[F.18: For "will-move-from" parameters, pass by `X&&` and `std::move` the parameter](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Rf-in)

with an exception
> Unique owner types that are move-only and cheap-to-move, such as unique_ptr, can also be passed by value which is simpler to write and achieves the same effect. Passing by value does generate one extra (cheap) move operation, but prefer simplicity and clarity first.

Reply:
Two suggestions from a brainstorming with @Dr15Jones 

20. Pass by value arguments which are not to be modiﬁed and are built-in types or small objects; otherwise pass arguments of class types by reference or, if necessary, by pointer. Allowed exception: if with a const-reference argument the function would make a copy of the argument, and the argument type is known to be moved more efficiently than copied (or the argument is move-only type), pass that argument by value. If the function is e.g. a member value setter, use `std::move` to set the member value. 

or 

20. Pass by value arguments which are not to be modiﬁed and are built-in types or small objects; otherwise pass arguments of class types by reference or, if necessary, by pointer. Allowed exception: if the function algorithm would benefit from using a move of the argument instead of a copy, pass the argument by value.

Either way, we'd clearly need examples.